File: plugins/operate/flow-mcp/src/domain/event.rs
//! Event schema. Each mutation produces an [`Event`] serialized as one JSONL
//! line (the append-only write-ahead record); the MCP surface reads the log back
//! via `list_events`.

use serde::{Deserialize, Serialize};

use super::ids::ItemId;
use super::model::{Status, WaitGate};

/// A single flow-state delta. `tag`-style serde so a JSONL line is
/// self-describing and a WS client can switch on `"kind"`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Event {
    /// An item was created or replaced.
    ItemUpserted {
        /// Affected item.
        id: ItemId,
        /// Display title.
        title: String,
    },
    /// An item's WAIT/GO gate changed.
    GateSet {
        /// Affected item.
        id: ItemId,
        /// New gate value.
        gate: WaitGate,
    },
    /// An item's carriage status advanced.
    StatusPosted {
        /// Affected item.
        id: ItemId,
        /// New status.
        status: Status,
    },
    /// Token spend was appended to an item.
    SpendAppended {
        /// Affected item.
        id: ItemId,
        /// Tokens added in this event.
        delta: u64,
        /// Cumulative total after this event.
        total: u64,
    },
    /// An item's model assignment changed.
    ModelSet {
        /// Affected item.
        id: ItemId,
        /// New resolved model.
        model: String,
    },
    /// A connection was added.
    ConnectionAdded {
        /// Dependent item.
        from: ItemId,
        /// Prerequisite item.
        to: ItemId,
    },
    /// A connection was removed.
    ConnectionRemoved {
        /// Dependent item.
        from: ItemId,
        /// Prerequisite item.
        to: ItemId,
    },
    /// A human comment was appended to an item's plan as an annotation
    /// (roadmap #4 comment loop).
    Annotated {
        /// Annotated item.
        id: ItemId,
        /// The comment text recorded against the item's plan.
        text: String,
    },
    /// A full re-draft of an item was requested, carrying the human's commentary
    /// (roadmap #4 rewrite loop). The actual re-draft is external orchestration;
    /// this records the request and the item's new draft number.
    RewriteRequested {
        /// Item to be re-drafted.
        id: ItemId,
        /// The commentary handed to the carriage agent for the re-draft.
        comment: String,
        /// The item's draft number after this request.
        draft: u32,
    },
    /// An orchestrator system message.
    SysMsg {
        /// Free-text message.
        text: String,
    },
}

impl Event {
    /// Serialize this event as a single JSONL line (no trailing newline).
    pub fn to_jsonl(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }

    /// Parse a single JSONL line back into an [`Event`].
    pub fn from_jsonl(line: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(line)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::ids::ItemId;

    fn id(s: &str) -> ItemId {
        ItemId::new(s).unwrap()
    }

    fn samples() -> Vec<Event> {
        vec![
            Event::ItemUpserted {
                id: id("a"),
                title: "A".into(),
            },
            Event::GateSet {
                id: id("a"),
                gate: WaitGate::Wait,
            },
            Event::StatusPosted {
                id: id("a"),
                status: Status::Doing,
            },
            Event::SpendAppended {
                id: id("a"),
                delta: 10,
                total: 10,
            },
            Event::ModelSet {
                id: id("a"),
                model: "claude-opus-4-8".into(),
            },
            Event::ConnectionAdded {
                from: id("a"),
                to: id("b"),
            },
            Event::ConnectionRemoved {
                from: id("a"),
                to: id("b"),
            },
            Event::Annotated {
                id: id("a"),
                text: "looks good".into(),
            },
            Event::RewriteRequested {
                id: id("a"),
                comment: "redo with X".into(),
                draft: 3,
            },
            Event::SysMsg {
                text: "hello".into(),
            },
        ]
    }

    #[test]
    fn every_variant_round_trips_through_jsonl() {
        for ev in samples() {
            let line = ev.to_jsonl().unwrap();
            assert!(!line.contains('\n'), "jsonl line must be single-line");
            let back = Event::from_jsonl(&line).unwrap();
            assert_eq!(back, ev);
        }
    }

    #[test]
    fn jsonl_is_self_describing_with_kind() {
        let ev = Event::SysMsg { text: "hi".into() };
        let line = ev.to_jsonl().unwrap();
        assert!(line.contains("\"kind\":\"sys_msg\""));
    }

    #[test]
    fn malformed_line_errors() {
        assert!(Event::from_jsonl("not json").is_err());
        assert!(Event::from_jsonl("{\"kind\":\"nope\"}").is_err());
    }
}

File: plugins/flow/flow-mcp/lib/flow_mcp/event.rb
# frozen_string_literal: true

require "json"
require_relative "errors"

module FlowMcp
  # The append-only event schema (EARS-FLOW-088). Each mutation produces one
  # `kind`-tagged JSON object, serialized as a single JSONL line. Events are
  # represented as plain string-keyed Hashes so they round-trip through JSON with
  # no bespoke (de)serialization and render directly in `list_events`.
  module Event
    KINDS = %w[
      item_upserted gate_set status_posted spend_appended model_set
      connection_added connection_removed annotated rewrite_requested sys_msg
    ].freeze

    module_function

    def item_upserted(id, title) = { "kind" => "item_upserted", "id" => id.to_s, "title" => title }
    def gate_set(id, gate) = { "kind" => "gate_set", "id" => id.to_s, "gate" => gate }
    def status_posted(id, status) = { "kind" => "status_posted", "id" => id.to_s, "status" => status }

    def spend_appended(id, delta, total)
      { "kind" => "spend_appended", "id" => id.to_s, "delta" => delta, "total" => total }
    end

    def model_set(id, model) = { "kind" => "model_set", "id" => id.to_s, "model" => model }
    def connection_added(from, to) = { "kind" => "connection_added", "from" => from.to_s, "to" => to.to_s }
    def connection_removed(from, to) = { "kind" => "connection_removed", "from" => from.to_s, "to" => to.to_s }
    def annotated(id, text) = { "kind" => "annotated", "id" => id.to_s, "text" => text }

    def rewrite_requested(id, comment, draft)
      { "kind" => "rewrite_requested", "id" => id.to_s, "comment" => comment, "draft" => draft }
    end

    def sys_msg(text) = { "kind" => "sys_msg", "text" => text }

    # Serialize one event as a single JSONL line (no trailing newline).
    def to_jsonl(event) = JSON.generate(event)

    # Parse one JSONL line back into an event hash. A line that is not a valid
    # event — malformed JSON OR a recognised object with an unknown `kind` —
    # raises, so replay aborts the open rather than silently dropping state
    # (EARS-FLOW-091).
    def from_jsonl(line)
      parsed =
        begin
          JSON.parse(line)
        rescue JSON::ParserError => e
          raise FlowMcp::Error, "malformed event line: #{e.message}"
        end
      unless parsed.is_a?(Hash) && KINDS.include?(parsed["kind"])
        raise FlowMcp::Error, "malformed event line: unknown kind #{parsed.is_a?(Hash) ? parsed['kind'].inspect : 'n/a'}"
      end
      parsed
    end
  end
end
